### PR TITLE
feat: complete destroy

### PR DESCRIPTION
**Pull request by verbanicm**
Open to suggestions on the output formatting. The integration workflow is not representative of what the end state will look like in production.

Actual flow:

```
init
  -> modified or "just plan" directories
  -> destroy or plan -destroy directories

plan
  business as usual for modified dirs

plan-destroy
  checkout main
  guardian plan with -destroy flag


PR MERGED

init
  -> modified or "just plan" directories
  -> destroy or plan -destroy directories

apply
  business as usual for modified dirs

apply-destroy
  checkout HEAD-1 (`github.events.pull_request.before`)
  guardian apply (parse destroy from GCS plan file metadata)
```

* Checking out main/head-1 is needed because we potentially deleted the directory or want to delete all existing resources.
* Splitting the jobs allows us to customize per operation (plan/destroy).
* Using GCS object metadata, we can determine if we are doing a destroy during the apply phase for informing the end user as needed, this wont affect the actual apply process, which is driven by the plan file contents.

### CODE REVIEW

File: pkg/commands/apply/apply.go [hunk 1]
```diff
@@ -463,12 +476,17 @@ func (c *ApplyCommand) downloadGuardianPlan(ctx context.Context, path string) (p
...
 
-	exitCode, ok := metadata["plan_exit_code"]
+	exitCode, ok := metadata[plan.MetaKeyExitCode]
 	if !ok {
 		return nil, "", fmt.Errorf("failed to determine plan exit code: %w", err)
 	}
 	planExitCode = exitCode
 
+	planOperation, ok := metadata[plan.MetaKeyOperation]
+	if ok && strings.ToLower(planOperation) == plan.OperationDestroy {
```
Comment (drevell) [inline pkg/commands/apply/apply.go:486]: Maybe https://pkg.go.dev/strings#EqualFold here rather than ToLower

File: pkg/commands/plan/plan.go [hunk 2]
```diff
@@ -313,14 +332,19 @@ func (c *PlanCommand) updateResultCommentForActions(ctx context.Context, startCo
...
+		destroyText = DestroyCommentText
+	}
+
+	msgBody := fmt.Sprintf("%s 🟦 No changes for dir: `%s` %s%s", CommentPrefix, c.childPath, c.gitHubLogURL, destroyText)
 
 	if result.hasChanges || resultErr != nil {
 		var comment strings.Builder
 		if resultErr != nil {
-			fmt.Fprintf(&comment, "%s 🟥 Failed for dir: `%s` %s\n\n<details>\n<summary>Error</summary>\n\n```\n\n%s\n```\n</details>", CommentPrefix, c.childPath, c.gitHubLogURL, resultErr)
+			fmt.Fprintf(&comment, "%s 🟥 Failed for dir: `%s` %s%s\n\n<details>\n<summary>Error</summary>\n\n```\n\n%s\n```\n</details>", CommentPrefix, c.childPath, c.gitHubLogURL, destroyText, resultErr)
```
Comment (drevell) [inline pkg/commands/plan/plan.go:345]: Could this be a helper function like `failureMessage(prefix, dir, childPath, url, destroyText string, err error)`? No strong feelings from me, but I see this duplicated around this codebase.

File: pkg/util/util.go [hunk 3]
```diff
@@ -26,6 +26,16 @@ import (
 	"golang.org/x/exp/maps"
 )
 
+// SliceContainsOnly checks that all values of a slice are a single value.
```
Comment (drevell) [inline pkg/util/util.go:29]: Consider adding another sentence like "Returns true when the input is an empty slice."

**Review by drevell — APPROVED** [review]

File: pkg/commands/plan/plan.go (hunk 2 above)
Comment (verbanicm) [reply pkg/commands/plan/plan.go:345]: 100%, but this is tech debt, I think we should implement a Reporter interface and have different ways of writing output. Ill create an issue.

File: pkg/commands/apply/apply.go (hunk 1 above)
Comment (verbanicm) [reply pkg/commands/apply/apply.go:486]: ah, nice. thanks!

File: pkg/util/util.go (hunk 3 above)
Comment (verbanicm) [reply pkg/util/util.go:29]: Hm, I would expect it to be false, updated code, test, and comment.

File: pkg/commands/plan/plan.go (hunk 2 above)
Comment (verbanicm) [reply pkg/commands/plan/plan.go:374]: Created https://github.com/abcxyz/guardian/issues/355

File: pkg/commands/workflows/plan_status_comments.go [hunk 4]
```diff
@@ -84,7 +85,7 @@ func (c *PlanStatusCommentCommand) Flags() *cli.FlagSet {
 		Usage:   "The Guardian init job result status.",
 	})
 
-	f.StringVar(&cli.StringVar{
+	f.StringSliceVar(&cli.StringSliceVar{
```
Comment (sethvargo) [inline pkg/commands/workflows/plan_status_comments.go:88]: Why this change?

File: pkg/terraform/plan.go [hunk 5]
```diff
@@ -45,6 +46,10 @@ func planArgsFromOptions(opts *PlanOptions) []string {
 		args = append(args, "-compact-warnings")
 	}
 
+	if util.PtrVal(opts.Destroy) {
```
Comment (sethvargo) [inline pkg/terraform/plan.go:49]: As a follow-up, we should switch to https://github.com/abcxyz/pkg/blob/main/pointer/pointer.go and drop the util package. Filed https://github.com/abcxyz/guardian/issues/356.

File: pkg/commands/apply/apply.go [hunk 6]
```diff
@@ -358,12 +361,17 @@ func (c *ApplyCommand) createStartCommentForActions(ctx context.Context) (*githu
 
 	c.Outf("Creating start comment")
 
+	destroyText := ""
+	if c.isDestroy {
+		destroyText = DestroyCommentText
+	}
+
```
Comment (sethvargo) [inline pkg/commands/apply/apply.go:368]: I see this same pattern repeated a number of times (thing, if destroy, thing = somethingelse). I've usually found that's a hint that using an interface (or separate command) is actually a better option.

File: pkg/terraform/plan.go (hunk 5 above)
Comment (verbanicm) [reply pkg/terraform/plan.go:49]: Ack, will change in a follow up (didnt know that existed)

File: pkg/commands/workflows/plan_status_comments.go (hunk 4 above)
Comment (verbanicm) [reply pkg/commands/workflows/plan_status_comments.go:88]: tl;dr - we will have multiple jobs (plan and plan_destroy) to account for having to checkout different branches (pr sha for plan, main for plan_destroy). This means we will have an end status comment based on multiple job results.

File: pkg/commands/apply/apply.go (hunk 6 above)
Comment (verbanicm) [reply pkg/commands/apply/apply.go:368]: I was planning on creating a reporter interface, so this info can be passed of to create different strategies to create output, ie github, stdout, some other thing. Will resolve this in a follow up.
